cmd/genkey.go

```diff
@@ -0,0 +1,128 @@
+package cmd
+
+import (
+	"fmt"
+	"github.com/gofrs/uuid"
+	"github.com/kairos-io/enki/pkg/config"
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
+	"k8s.io/mount-utils"
+	"os"
+	"os/exec"
+	"path/filepath"
+)
+
+func NewGenkeyCmd() *cobra.Command {
+	c := &cobra.Command{
+		Use:   "genkey NAME",
+		Short: "Generate secureboot keys under the uuid generated by NAME",
+		Args:  cobra.ExactArgs(1),
+		RunE: func(cobraCmd *cobra.Command, args []string) error {
+			// Set this after parsing of the flags, so it fails on parsing and prints usage properly
+			cobraCmd.SilenceUsage = true
+			cobraCmd.SilenceErrors = true // Do not propagate errors down the line, we control them
+
+			cfg, err := config.ReadConfigBuild(viper.GetString("config-dir"), cobraCmd.Flags(), &mount.FakeMounter{})
+			if err != nil {
+				return err
+			}
+			l := cfg.Logger
+			uid := uuid.NewV5(uuid.NamespaceDNS, args[0])
+			output, _ := cobraCmd.Flags().GetString("output")
+
+			err = os.MkdirAll(output, 0755)
+			if err != nil {
+				l.Errorf("Error creating output directory: %s", err)
+				return err
+			}
+
+			for _, t := range []string{"PK", "KEK", "DB"} {
+				l.Infof("Generating %s", t)
+				key := filepath.Join(output, fmt.Sprintf("%s.key", t))
+				pem := filepath.Join(output, fmt.Sprintf("%s.pem", t))
+				der := filepath.Join(output, fmt.Sprintf("%s.der", t))
+				auth := filepath.Join(output, fmt.Sprintf("%s.auth", t))
+				esl := filepath.Join(output, fmt.Sprintf("%s.esl", t))
+				cmd := exec.Command(
+					"openssl",
+					"req", "-nodes", "-x509", "-subj", fmt.Sprintf("/CN=%s/", args[0]),
+					"-keyout", key,
+					"-out", pem,
+				)
+				out, err := cmd.CombinedOutput()
+				if err != nil {
+					l.Errorf("Error generating %s: %s", t, string(out))
+					return err
+				}
+				l.Infof("%s generated at %s and %s", t, key, pem)
+
+				l.Infof("Converting %s.pem to DER", t)
+				cmd = exec.Command(
+					"openssl", "x509", "-outform", "DER", "-in", pem, "-out", der,
+				)
+				out, err = cmd.CombinedOutput()
+				if err != nil {
+					l.Errorf("Error generating %s: %s", t, string(out))
+					return err
+				}
+				l.Infof("%s generated at %s", t, der)
+
+				l.Infof("Generating %s.esl", t)
+				cmd = exec.Command(
+					"sbsiglist", "--owner", uid.String(), "--type", "x509", "--output", esl, der,
+				)
+				out, err = cmd.CombinedOutput()
+				if err != nil {
+					l.Errorf("Error generating %s: %s", t, string(out))
+					return err
+				}
+				l.Infof("%s generated at %s", t, esl)
+
+				// For PK and KEK we use PK.key and PK.pem to sign it
+				// For DB we use KEK.key and KEK.pem to sign it
```

Comment:
😅 that is confusing if KEK use PK otherwhise use KEK, LOL. Glad you added a comment